brainlit/algorithms/generate_fragments/state_generation.py

```diff
@@ -439,6 +441,45 @@ def _endpoints_from_coords_neighbors(self, coords):
 
         return ends
 
+    def _pc_endpoints_from_coords_neighbors(self, coords):
+        """Compute endpoints of fragment with Principal Curves.
+
+        Args:
+            coords (np.array): coordinates of voxels in the fragment
+
+        Returns:
+            list: endpoints of the fragment
+
+        References
+        ----------
+        .. [1] Hastie, Trevor, and Werner Stuetzle. “Principal Curves.”
+        Journal of the American Statistical Association, vol. 84, no. 406,
+        [American Statistical Association, Taylor & Francis, Ltd.], 1989,
+        pp. 502–16, https://doi.org/10.2307/2289936.
+        .. [2] Principal Curves Code written by zsteve,
+        https://github.com/zsteve
+        https://github.com/zsteve/pcurvepy
+        """
+
+        ends = []
+
+        # Make sure x, y, z ascending & don't repeat
+        sorter = np.lexsort((coords[:, 2], coords[:, 1], coords[:, 0]))
+        coords = coords[sorter]
+        coords = np.unique(coords, axis=0)
+
+        p_curve = pcurve.PrincipalCurve(k=1, s_factor=5)
+        p_curve.fit(coords, max_iter=50)
+        pc = p_curve.p
+
+        pc = np.floor(pc + 0.5)
+        pc_frag_list = [i for i in pc if i in coords]
+
+        ends.append(pc_frag_list[0])
```

Comment:
can you remind me, did we agree that the best thing to do is take the first and last point? are we sure this will be a voxel that is part of the fragment?

Reply:
test

Reply:
@tathey1 You're right, we discussed making sure the estimated endpoint voxels were actually part of the fragment. The line `pc_frag_list = [i for i in pc if i in coords]` guarantees the endpoint candidates are part of the fragment. Then, it is best to take the fragment voxels that appear earliest and furthest along the principal curve respectively. This is why I have `ends.append(pc_frag_list[0])` and `ends.append(pc_frag_list[-1])`